src/odemis/odemisd/main.py

```diff
@@ -321,78 +322,95 @@ def _terminate_all_alive(self):
         """
         Stops all the components that are currently alive (excepted the
           microscope)
-        It terminate the parents (aka "users") first as the instantiated children should
-         never need their parent but the parent might rely on the children.
-         Delegated children should be directly terminated by their creator, so
-         by the time we terminate them, it's a no-op.
+        It terminates the dependents (aka "users") first as the dependencies should
+         never need their dependent but the dependent might rely on the dependency.
+         Children will be terminated before their creator.
         It also stops the containers, once no component is running in them.
         """
         mic = self._instantiator.microscope
-        alive = set(mic.alive.value)
-
-        # create a "graph" child->parents
-        parents = {c: set() for c in alive}  # comp -> set of comps (all its parents)
-        for comp in alive:
-            # TODO: if a component was created by delegation, use its creator
-            # instead of itself as "used" by the other components, to ensure
-            # they might not be terminated too early.
+
+        # alive component -> components which depends on it
+        self._dependents = {c: set() for c in set(mic.alive.value)}  # comp -> set of comps (all the comps that depend on it)
+        # Children will be terminated when their parents are terminated, so no need to keep them in .dependents
+        for comp in list(self._dependents.keys()):  # ._dependents will be modified, so we need a copy
+            for child in comp.children.value:
+                self._dependents.pop(child, None)
+        for comp in self._dependents:
             try:
-                for child in comp.children.value:
-                    # TODO: do not add the creator
-                    parents[child].add(comp)
+                deps = comp.dependencies.value
             except Exception:
```

Comment:
Do we really need the try-except here? If so, we should probably also do it in line 336 with `comp.children`.